### ISSUE TITLE
In-Game: Decouple audio-decoding threads from the audio context

### ISSUE THREAD

**Issue by toby-yoyo** [YoYoGames/GameMaker-Bugs#7055]
### Is your feature request related to a problem?

The primary aims of the audio decoding threads are:
- To decode a given source of compressed audio data.
- To seek around the source when requested.
- To provide some information about the codec state (stream position etc.).

Our decoding threads do achieve these things. However, they also do some other things, such as:
- Buffering decoded audio data.
- Queueing/Unqueueing buffers on the audio context.
- Getting/Setting source states.

The result of this is that many simple processes in the audio engine (e.g. stopping a sound) become an async process where we have to defer actions to the decoding threads. On top of this, we need to present these async processes as synchronous to users, so we have to build up rather complex systems to achieve this.

Additionally, the decoding threads have a few problematic behaviours such as 'detecting' buffer underruns, leading to potentially incorrect source state overrides, and partial deinitialisation of voice subsystems, which steps on the feet of the main thread. There is also a tight coupling of the stream states and source states which doesn't make much sense as these are two mostly unrelated systems.

Finally, we have at minimum 6 different threads that access the audio context (plus more depending on other features that users may/may not be utilising). This presents problems such as increased contention for the context (which is handled by only a single mutex), and loss of error state (which is intended to be read from a single thread).

### Describe the solution you'd like

The decoding threads already do the main things they are supposed to, and generally in a reasonable way as well. However, moving the responsibility of certain tasks to the main/audio threads will help to untangle the code somewhat.

We should decouple the decoding threads entirely from the audio context. Decoding threads should not really need to know about or interact with the audio context, or even the engine more generally as they are essentially a specialised file/stream reader. 

Specific changes we would make here would include:
- Deferring buffering and queueing of data to the audio thread.
- Consolidation of various distributed state flags into a concrete set of voice states, including an initial 'preparing' and final 'ramping' state.
- Allowing the audio thread to make certain voice state transitions (e.g. preparing->playing and ramping->stopped).
- Enabling voices to detach from decoding tasks, allowing them to be reused without waiting for the decoding task to deinitialise.
- Wrapping the decoding threads/tasks in a generic interface to allow us to integrate new codecs more easily.
- Implementing OpenAL's source types, which provide information about how a source's buffers are provided.

### Additional context

This isn't a huge change, but it should help us towards cleaning up a more awkward aspect of our voice handling. Improvements to voice handling more generally will help us to pave the way for features that rely heavily on voice management, such as audio scheduling. This in turn will allow us to address issues in existing features such as audio queues and sync groups, which can only be used in limited ways and are mostly isolated from the other parts of the audio engine.

**Comment by toby-yoyo** [comment]
Various things have changed due to this task:

### `audio_channel_num`
- This function will respect the number passed into it.
- An internal lower limit of voices has now been removed, so users can now pass 0 to the function to disable audio playback.
- Channels used to play sync groups are now part of the pool that is resized by this function.
- Note that audio from video playback will not be affected by this function on most platforms.

### `audio_debug`
- Two new source states will be displayed in the audio debug window:
    - 'Waiting' sources are preparing to play and will begin playback once some internal conditions have been satisfied. These are coloured in magenta.
    - 'Fading' sources will stop after the next mix occurs on the audio thread. These are coloured in cyan.
- Sync group sounds will be visible in this window (see below).
- The `numProcessed` column will now always display zero. We may repurpose this column to show which source is being synced to (for sync groups).

### Audio sync groups
- Sounds in sync groups will now each have their own source. Previously, they were premixed and routed to a single source.
- Sources assigned to sounds in sync groups will no longer be separate from the main voice pool, and so they will be affected by the same rules and behaviours that regular sounds are (voice stealing, `audio_stop_all` etc.). Note that sync group sounds are currently played with the maximum possible priority.
- Because of the above, sounds in sync groups can now be debugged like other sounds using `audio_debug` window. This deprecates `audio_sync_group_debug` which will be stubbed going forward.
- Sounds in sync groups have had certain restrictions lifted:
    - Sounds no longer need to have the same sample rate.
    - Sounds no longer need to have the same channel format.
    - Sounds no longer need to be exclusively compressed - a mixture of compressed and uncompressed sounds can be played together in a sync group.
- Sync groups will now be supported on the GX.games target.

**Comment by gurpreetsinghmatharoo** [comment]
Documentation updated as per Toby's comment

**Comment by alicemoretti** [comment]
On Beta IDE v2024.1100.0.626  Runtime v2024.1100.0.652, the project attached crashes on HTML5. 

Reopening the issue. Thank you.

**Comment by toby-yoyo** [comment]
Apologies, I should have mentioned that this is a C++ runner-only change and doesn't affect the HTML5 runner as they have two entirely different audio systems. The project wasn't really written in HTML5 in mind.

That being said, the project did actually expose a few bugs on HTML5, which I've raised the following issues for:
- https://github.com/YoYoGames/GameMaker-Bugs/issues/7804
- https://github.com/YoYoGames/GameMaker-Bugs/issues/7805
- https://github.com/YoYoGames/GameMaker-Bugs/issues/7806

I've also edited the project to be a bit more HTML5-friendly, which can be accessed from my comment above. With some quick-fixes in for #7804 and #7805 on my branch, the project runs the whole way through without crashing, though it does complain about #7806 which we should fix too.

Re-closing this one as the HTML5 issues are unrelated to the task and have now been logged separately. Thank you for raising that.
